source/common/upstream/upstream_connection_pool_impl.cc

```diff
@@ -0,0 +1,173 @@
+  dispatcher_.clearDeferredDeleteList();
+}
+
+void UpstreamConnectionPoolImpl::OfferConnection(HostConstSharedPtr host,
+                                                 UpstreamConnectionEssence connection_essence) {
+  auto shared_connection_essence =
+      std::make_shared<UpstreamConnectionEssence>(std::move(connection_essence));
+  dispatcher_.post([this, host, shared_connection_essence]() {
+    AcceptConnectionInDispatcherThread(host, std::move(*shared_connection_essence));
+  });
+  {
+    Thread::LockGuard lock(mu_);
+    ConnectionPool& pool = connection_pools_[host];
+    ++pool.available_connections;
+  }
+}
+
+bool UpstreamConnectionPoolImpl::RetrieveConnection(
+    HostConstSharedPtr host, Event::Dispatcher& dispatcher,
+    std::function<void(UpstreamConnectionEssence connection_essence)> accept_cb) {
+  {
+    Thread::LockGuard lock(mu_);
+    ConnectionPool& pool = connection_pools_[host];
+    if (pool.available_connections <= 0) {
+      return false;
+    }
+    --pool.available_connections;
+  }
+  dispatcher_.post([this, host, &dispatcher, accept_cb]() {
+    RetrieveConnectionInDispatcherThread(host, dispatcher, accept_cb);
+  });
+  return true;
+}
+
+void UpstreamConnectionPoolImpl::AcceptConnectionInDispatcherThread(
+    HostConstSharedPtr host, UpstreamConnectionEssence connection_essence) {
+  Thread::LockGuard lock(mu_);
+  ConnectionPool& pool = connection_pools_[host];
+
+  auto client_connection = dispatcher_.adoptClientConnection(
+      std::move(connection_essence.socket), std::move(connection_essence.transport_socket));
+  ActiveClientPtr new_client(new ActiveClient(*this, host, std::move(client_connection),
+                                              connection_essence.remaining_requests));
+  new_client->moveIntoList(std::move(new_client), pool.clients);
+}
+
+void UpstreamConnectionPoolImpl::RetrieveConnectionInDispatcherThread(
+    HostConstSharedPtr host, Event::Dispatcher& dispatcher,
+    std::function<void(UpstreamConnectionEssence connection_essence)> accept_cb) {
+  Thread::LockGuard lock(mu_);
+  ConnectionPool& pool = connection_pools_[host];
+  if (pool.clients.empty()) {
+    // We're over-committed on number of available pooled connections. Release our lease and fail
+    // the connection pooling attempt.
+    ++pool.available_connections;
+    dispatcher.post([accept_cb]() {
+      UpstreamConnectionEssence no_connection;
+      accept_cb(std::move(no_connection));
+    });
+
+    return;
+  }
+  ActiveClientPtr client = std::move(pool.clients.front());
+  pool.clients.pop_front();
+  auto shared_connection_essence =
+      std::make_shared<UpstreamConnectionEssence>(client->detachSockets());
+  client.reset();
+  dispatcher.post([accept_cb, shared_connection_essence]() {
+    accept_cb(std::move(*shared_connection_essence));
+  });
+}
+
+void UpstreamConnectionPoolImpl::onEvent(ActiveClient& client, Network::ConnectionEvent /*event*/) {
```

Comment:
So assuming that this is a connection close is fine because we don't hand over a connection to the shared pool until it has already been connected? Probably warrants a comment